### ISSUE TITLE
Overriding association methods

### ISSUE THREAD

**Issue by mateomurphy** [rails-api/active_model_serializers#848]
While going through the tests, I noticed a behaviour that seems wrong to me, but I want to validate before submitting a fix:

We have a serializer:

```
PostSerializer = Class.new(ActiveModel::Serializer) do
  cache key:'post', expires_in: 0.05
  attributes :id, :title, :body

  has_many :comments
  belongs_to :blog
  belongs_to :author
  url :comments

  def blog
    Blog.new(id: 999, name: "Custom blog")
  end

  def custom_options
    options
  end
end
```

And the following test setup (edited for clarity)

```
@blog = Blog.new(id: 23, name: 'AMS Blog')
@first_post = Post.new(id: 1, title: 'Hello!!', body: 'Hello, world!!')
@second_post = Post.new(id: 2, title: 'New Post', body: 'Body')
@first_post.blog = @blog
@second_post.blog = nil
```

And that gives the following output

```
{ title: "Hello!!", body: "Hello, world!!", id: "1", links: { comments: [], blog: "999", author: "1" } },
{ title: "New Post", body: "Body", id: "2", links: { comments: [], blog: nil, author: "1" } }
```

Now, with the overridden association, I would expect it to always return blog: "999", but it only does so if Post has a value for blog. Is this a bug?

**Comment by JoshSmith** [comment]
What version are you on?

**Comment by mateomurphy** [comment]
@JoshSmith  latest master: the test code I've pasted above is from here https://github.com/rails-api/active_model_serializers/blob/master/test/adapter/json_api/collection_test.rb

**Comment by JoshSmith** [comment]
Thanks!

**Comment by kurko** [comment]
@mateomurphy yes. Likely in `serializer#each_association`. Would you venture a PR? :smile:

**Comment by mateomurphy** [comment]
Here: https://github.com/rails-api/active_model_serializers/pull/850

**Comment by mateomurphy** [comment]
Found another bug in the same method https://github.com/rails-api/active_model_serializers/pull/852

**Comment by joaomdmoura** [comment]
@mateomurphy :+1: keep up the good work!

**Comment by JoshSmith** [comment]
Closed by #850. Also excellent work in #852.
